book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -0,0 +1,684 @@
+
+1. Check if the parent block exists in any current chain
+
+1. If it does, call `let ret = self.commit_block(block)`
+    - Call `self.process_queued(new_parents)` if `ret` is `Some`
+
+1. Else Add `block` to `self.queued_blocks` and related members and return
+
+### `fn process_queued(&mut self, new_parent: block::Hash)`
+
+1. Create a list of `new_parents` and populate it with `new_parent`
+
+1. While let Some(parent) = new_parents.pop()
+    - for each `hash` in `self.queued_by_parent.remove(&parent.hash)`
+      - lookup the `block` for `hash`
+      - remove `block` from `self.queued_blocks`
+      - remove `hash` from `self.queued_by_height`
+      - let result = `self.commit_block(block)`;
+      - add `result` to `new_parents`
+
+### `fn commit_block(&mut self, block: QueuedBlock) -> Option<block::Hash>`
+
+Try to commit `block` to the non-finalized state. Returns `None` if the block
+cannot be committed due to missing context.
+
+1. For each `chain`
+    - if `block.parent` == `chain.tip`
+      - try to push `block` onto that chain
+      - broadcast `result` via `block.rsp_tx`
+      - return Some(block.hash) if `result.is_ok()`
+
+1. Find the first chain that contains `block.parent` and fork it with
+  `block.parent` as the new tip
+    - `let fork = self.chains.iter().find_map(|chain| chain.fork(block.parent));`
+
+1. If `fork` is `Some`
+    - try to push `block` onto that chain
+      - if successful add `fork` to `self.chains`
+    - broadcast `result` via `block.rsp_tx`
+    - return Some(block.hash) if `result.is_ok()`
+
+1. Else panic, this should be unreachable because `commit_block` is only
+   called when it's ready to be committed.
+
+### Summary
+
+- `Chain` represents the non-finalized portion of a single chain
+
+- `ChainSet` represents the non-finalized portion of all chains and all
+  unverified blocks that are waiting for context to be available.
+
+- `chain_set::queue` handles queueing and or commiting blocks and
+  reorganizing chains (via `commit_block`) but not finalizing them
+
+- Finalized blocks are returned from `finalize` and must still be committed
+  to disk afterwards
+
+- `finalize` handles pruning queued blocks that are past the reorg limit
+
+## Committing non-finalized blocks
+
+Given the above structures for manipulating the non-finalized state new
+`non-finalized` blocks are commited in two steps. First we commit the block
+to the in memory state, then we finalize the lowest height block if it is
+past the reorg limit, finally we process any queued blocks and prune any that
+are now past the reorg limit.
+
+1. Try to commit or queue the block to the non-finalized state with
+   `chain_set.queue(block)?;`
+
+1. If the best chain is longer than the reorg limit
+    - Finalize the lowest height block in the best chain with
+     `let finalized = chain_set.finalize()?;`
+    - commit `finalized` to disk with `CommitFinalizedBlock`
```

Comment:
```suggestion
    - Finalize all lowest height blocks in the best chain, and commit them to disk with `CommitFinalizedBlock`:
      ```
      while let finalized = chain_set.finalize()? {
        let request = CommitFinalizedBlock { finalized };
        sled_state.ready_and().await?.call(request).await?;
      };
      ```
```

Reply:
I don't think we should do a `while let` with `finalize`, I'd like to keep all the control flow logic in one place, so finalize always unconditionally finalizes the next block while this top level commit logic handles checking the height of the best chain to see if it needs to call finalize.

```suggestion
    - Finalize all lowest height blocks in the best chain, and commit them to disk with `CommitFinalizedBlock`:
      ```
      while self.best_chain().len() > reorg_limit {
        let finalized = chain_set.finalize()?;
        let request = CommitFinalizedBlock { finalized };
        sled_state.ready_and().await?.call(request).await?;
      };
      ```
```